### ISSUE TITLE
Send emails on charitable-completed_donation status instead of save_post

### ISSUE THREAD

**Issue by helgatheviking** [Charitable/Charitable#217]
Suggest that:
emails support rich text as well as plain text
emails use templates with hooks
shortcode callbacks are templates in lieu of getting strings via filters. 

ex: `[charitable_email show=donation_summary]` calls a template

Edit: Most importantly, their needs to be a custom trigger method! Right now, every email is triggered by `save_post` but it would be more flexible to have the emails triggered by specific hooks... some kind of hook for changing status might be best. When a donation is marked as paid, that could trigger the receipt email, etc.

**Comment by ericnicolaas** [comment]
@helgatheviking Just circling around to this one. A few thoughts: 

- Emails already support rich text. Do you mean for emails to be sent as multi-part email? 
- Using templates instead of filters might work for some of the shortcode options (like `donation_summary`), but there are many for which it doesn't make sense. For example, things like `donor_first_name`, `donor_email`, `donation_id`, `campaign_title`, etc. kind of just make sense as having a string returning the value. 
- What do you mean by "templates with hooks"?
- I think switching the hook that fires off the donation receipt & admin notification makes sense as a `charitable-completed_donation` (via [`wp_transition_post_status`](https://developer.wordpress.org/reference/functions/wp_transition_post_status/)).

Let's also create separate issues for any of these that we decide to tackle, just so each issue is focused on a single problem.

**Comment by helgatheviking** [comment]
I created this issue when I was dead set on putting the subscription info on the *same* email as the donation receipt. That turned out to be verrrry difficult. But, now that we've moved to creating new emails for recurring donations this is maybe less important.

As for templates, I was thinking something like: 

```
// Example Receipt Email
do_action('charitable_email_header', $email, $donation_id );
do_action('charitable_email_content', $email, $donation_id );
do_action('charitable_email_footer', $email, $donation_id );

```

That might be easier to work with than the current

		protected function build_email() {
			ob_start();

			charitable_template( 'emails/header.php', array( 'email' => $this ) );

			charitable_template( 'emails/body.php', array( 'email' => $this ) );

			charitable_template( 'emails/footer.php', array( 'email' => $this ) );

			$message = ob_get_clean();

			return apply_filters( 'charitable_email_message', $message, $this );
		}

You could move things around, or add new content before/after via priority. Still might be nice, but like I said, less important for me moving forward. 

Changing the hook that fires the email remains important though, atleast imho.

**Comment by helgatheviking** [comment]
Definitely need to change [this](https://github.com/charitable/Charitable/blob/dev/includes/emails/charitable-email-hooks.php#L30-L31)

```
add_action( 'charitable_after_save_donation', array( 'Charitable_Email_Donation_Receipt', 'send_with_donation_id' ) );
add_action( 'save_post_' . Charitable::DONATION_POST_TYPE, array( 'Charitable_Email_Donation_Receipt', 'send_with_donation_id' ) );
```

to 

```
add_action( 'charitable-completed_' . Charitable::DONATION_POST_TYPE, array( 'Charitable_Email_Donation_Receipt', 'send_with_donation_id' ) );
```

so that I can remove that action when the donation is part of a subscription

**Comment by ericnicolaas** [comment]
OK, I'll make sure I update the hook. 

I'm not convinced about the value of shifting to the `do_action` calls for loading the parts of the email. You'd still need to load those templates, so you'd probably end up doing the same thing in a more roundabout way.

**Comment by helgatheviking** [comment]
errr.... I may need to rethink the 

```
add_action( $status . '_' . Charitable::DONATION_POST_TYPE, array( 'Charitable_Email_Donation_Receipt', 'send_with_donation_id' ) );
```

because if you create a donation with an initial status of `charitable-completed` then the email is triggered on `wp_insert_post()` which is too early because it is *before* the processor gets to run `save_campaign_donations()`. So then `send_with_donation_id()` tries to run but it can't find the campaign donations and so can't get an email address from a user. 

I'm seeing this when creating renewal donations when the PayPal IPN sends a new recurrence, but also if I just try to test:

```
$campaign_id = 139; // ID of campaign

$args = array(            
    'status' => 'charitable-completed', 
    'gateway' => 'manual', 
    'campaigns' => array(
        array( 
            'campaign_id' => $campaign_id,
            'amount' => 50,
        )
    ), 
    'user' => array(
        'first_name' => 'Kara',
        'last_name' => 'Thrace',
        'email' => 'starbuck@example.com'
    )
);
```

here's the error trace:
> [22-Sep-2016 16:11:58 UTC] PHP Notice:  Trying to get property of non-object in /home/ubuntu/workspace/wp-content/plugins/charitable/includes/donations/abstract-charitable-donation.php on line 409
> [22-Sep-2016 16:11:58 UTC] PHP Stack trace:
> [22-Sep-2016 16:11:58 UTC] PHP   1. {main}() /home/ubuntu/workspace/wp-admin/admin-ajax.php:0
> [22-Sep-2016 16:11:58 UTC] PHP   2. require_once() /home/ubuntu/workspace/wp-admin/admin-ajax.php:22
> [22-Sep-2016 16:11:58 UTC] PHP   3. require_once() /home/ubuntu/workspace/wp-load.php:39
> [22-Sep-2016 16:11:58 UTC] PHP   4. require_once() /home/ubuntu/workspace/wp-config.php:104
> [22-Sep-2016 16:11:58 UTC] PHP   5. include() /home/ubuntu/workspace/wp-settings.php:385
> [22-Sep-2016 16:11:58 UTC] PHP   6. charitable_create_donation($args = *uninitialized*) /home/ubuntu/workspace/wp-content/themes/nonprofit/functions.php:215
> [22-Sep-2016 16:11:58 UTC] PHP   7. Charitable_Donation_Processor->save_donation($values = *uninitialized*) /home/ubuntu/workspace/wp-content/plugins/charitable/includes/donations/charitable-donation-functions.php:64
> [22-Sep-2016 16:11:58 UTC] PHP   8. wp_insert_post($postarr = *uninitialized*, $wp_error = *uninitialized*) /home/ubuntu/workspace/wp-content/plugins/charitable/includes/donations/class-charitable-donation-processor.php:385
> [22-Sep-2016 16:11:58 UTC] PHP   9. wp_transition_post_status($new_status = *uninitialized*, $old_status = *uninitialized*, $post = *uninitialized*) /home/ubuntu/workspace/wp-includes/post.php:3324
> [22-Sep-2016 16:11:58 UTC] PHP  10. do_action($tag = *uninitialized*, $arg = *uninitialized*, *uninitialized*) /home/ubuntu/workspace/wp-includes/post.php:3897
> [22-Sep-2016 16:11:58 UTC] PHP  11. call_user_func_array:{/home/ubuntu/workspace/wp-includes/plugin.php:524}(*uninitialized*, *uninitialized*) /home/ubuntu/workspace/wp-includes/plugin.php:524
> [22-Sep-2016 16:11:58 UTC] PHP  12. Charitable_Email_Donation_Receipt::send_with_donation_id($donation_id = *uninitialized*) /home/ubuntu/workspace/wp-includes/plugin.php:524
> [22-Sep-2016 16:11:58 UTC] PHP  13. Charitable_Email->send() /home/ubuntu/workspace/wp-content/plugins/charitable/includes/emails/class-charitable-email-donation-receipt.php:89
> [22-Sep-2016 16:11:58 UTC] PHP  14. Charitable_Email_Donation_Receipt->get_recipient() /home/ubuntu/workspace/wp-content/plugins/charitable/includes/emails/abstract-class-charitable-email.php:983
> [22-Sep-2016 16:11:58 UTC] PHP  15. Charitable_Abstract_Donation->get_donor() /home/ubuntu/workspace/wp-content/plugins/charitable/includes/emails/class-charitable-email-donation-receipt.php:113
> [22-Sep-2016 16:11:58 UTC] PHP  16. Charitable_Abstract_Donation->get_donor_id() /home/ubuntu/workspace/wp-content/plugins/charitable/includes/donations/abstract-charitable-donation.php:421

**Comment by helgatheviking** [comment]
I *think* the solution is to leave the `send_with_donation_id()` attached to both the post status and the `charitable_after_save_donation` hook. And to modify `send_with_donation_id()` to include something like:

$donation = charitable_get_donation( $donation_id );
            
        if ( ! is_object( $donation ) || empty( $donation->get_campaign_donations() ) || ! apply_filters( 'charitable_send_' . self::get_email_id(), true, $donation ) ){
            return false;
        }

Specifically, the `empty( $donation->get_campaign_donations() )` will prevent the email from firing in the instance where `charitable-completed_donation` is called before the full donation is completed.... ie: if inserting a donation directly with the `charitable-completed` status.

**Comment by ericnicolaas** [comment]
Closed in 19866e979f70ebb10a130b03df1128e0d372ccb6
